ghc/Categorifier/GHC/Core.hs

```diff
@@ -0,0 +1,381 @@
+--   simplifier does.
+--
+--   Here we try to give more control over how the simplifier is applied, making it possible to at
+--   least call `simplifyExpr'` with a `SimplMode`. We also try to provide access to specific
+--   transformations so that Core plugins can do fine-grained simplification internally.
+--
+--  __NB__: Unfortunately, many of the useful operations aren't exposed in the API, so this
+--          duplicates them as necessary, with comments explaining which are abstracted (so we can
+--          try to keep them in sync with upstream changes).
+--
+-- | Like `simplifyExpr`, except it takes a `Set` of `Transformation`s to apply. This
+--   allows us to control (as much as possible) how the simplifier is applied in each call.
+--
+--   No matter how we restrict it, this simplification applies a ton of transformations that are
+--   hard to separate out (future work?). This one disables everything we can by default, then
+--   allows us to turn on individual transformations. There are also some other transformations we
+--   need that are implicit in here. E.g. specialization and case of known constructor.
+simplifyExpr' ::
+  Utils.Logger ->
+  Driver.DynFlags ->
+  Set Transformation ->
+  Types.UniqSupply ->
+  CoreExpr ->
+  IO CoreExpr
+#if MIN_VERSION_ghc(9, 2, 0)
+simplifyExpr' logger dflags trans _ expr =
+  fmap fst . initSmpl logger dflags emptyRuleEnv emptyFamInstEnvs (exprSize expr) $
+    doSimplify
+      1
+      SimplMode
+        { sm_case_case = CaseOfCase `Set.member` trans,
+          sm_uf_opts = defaultUnfoldingOpts,
+          sm_pre_inline = Inline `Set.member` trans,
+          sm_logger = logger,
+          sm_dflags = dflags,
+          sm_eta_expand = EtaExpand `Set.member` trans,
+          sm_inline = Inline `Set.member` trans,
+          sm_names = ["categorify internal"],
+          sm_phase = Types.Phase 1,
+          sm_rules = Rules `Set.member` trans -- this improves specialisation
+        }
+      expr
+#else
+simplifyExpr' _ dflags trans uniqS expr =
+  fmap fst . initSmpl dflags emptyRuleEnv emptyFamInstEnvs uniqS (exprSize expr) $
+    doSimplify
+      1
+      SimplMode
+        { sm_case_case = CaseOfCase `Set.member` trans,
+          sm_dflags = dflags,
+          sm_eta_expand = EtaExpand `Set.member` trans,
+          sm_inline = Inline `Set.member` trans,
+          sm_names = ["categorify internal"],
+          sm_phase = Types.Phase 1,
+          sm_rules = Rules `Set.member` trans -- this improves specialisation
+        }
+      expr
+#endif
+
+-- | Designed to be like `CoreDoSimplify`, but applicable to arbitrary `CoreExpr`s.
+doSimplify ::
+  Int ->
+  SimplMode ->
+  CoreExpr ->
+  SimplM CoreExpr
+doSimplify passCount mode expr =
+  foldr (<=<) pure (replicate passCount $ simplExpr simplEnv . occurAnalyseExpr) expr
+  where
+    simplEnv' = mkSimplEnv mode
+    simplEnv =
+      simplEnv' {seInScope = Types.extendInScopeSetSet (seInScope simplEnv') $ exprFreeVars expr}
```

Comment:
Please add a comment here explaining why adding `exprFreeVars expr` to `seInScope`

Reply:
Really good point.